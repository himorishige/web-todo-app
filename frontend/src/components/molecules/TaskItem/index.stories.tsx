/** @jsxImportSource @emotion/react */
import { ComponentStory, ComponentMeta } from '@storybook/react';
import TaskItem from '.';

export default {
  title: 'WebToDoApp/Molecules/TaskItem',
  component: TaskItem,
} as ComponentMeta<typeof TaskItem>;

const Template: ComponentStory<typeof TaskItem> = (args) => <TaskItem {...args} />;

export const Default = Template.bind({});
Default.storyName = 'TaskItem';
