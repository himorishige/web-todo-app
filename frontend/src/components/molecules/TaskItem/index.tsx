/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import { Link } from 'react-router-dom';
import { CheckBox, StarIcon } from 'src/components/atoms';

type Props = {
  id: string;
  title: string;
  isCompleted: boolean;
  priority: number;
  completedStateHandler: (id: string, isCompleted: boolean) => Promise<void>;
  priorityStateHandler: (id: string, priority: number) => Promise<void>;
};

const TaskItem: React.VFC<Props> = (props) => {
  const { id, title, isCompleted, priority, completedStateHandler, priorityStateHandler } = props;

  return (
    <div css={taskItemStyle}>
      <div css={itemLabelStyle}>
        <CheckBox
          isCompleted={isCompleted}
          onChange={() => completedStateHandler(id, !isCompleted)}
        />
        <div css={labelTextStyle}>
          <Link to={`/tasks/${id}`}>{title}</Link>
        </div>
      </div>
      <div>
        <StarIcon status={priority} onClick={() => priorityStateHandler(id, +!priority)} />
      </div>
    </div>
  );
};

export default TaskItem;

const taskItemStyle = css`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  padding: 1.25rem;
  background: #f8f8f8;
  border-radius: 4px;
`;

const itemLabelStyle = css`
  display: flex;
  width: 100%;
  align-items: center;
  label {
    flex-basis: 3rem;
  }
`;

const labelTextStyle = css`
  width: 100%;
  a {
    display: block;
    width: 100%;
    color: var(--black-color);
    text-decoration: none;
    &:hover {
      text-decoration: underline;
      color: var(--secondary-color);
    }
  }
`;
