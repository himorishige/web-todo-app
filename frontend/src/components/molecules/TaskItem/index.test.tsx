import { cleanup, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import TaskItem from '.';
import { matchers } from '@emotion/jest';
import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

// Add the custom matchers provided by '@emotion/jest'
expect.extend(matchers);

const id = '1';
const title = 'タイトル';
const description = '説明文';
const isCompleted = false;
const priority = 0;
const completedStateHandler = jest.fn();
const priorityStateHandler = jest.fn();

afterEach(() => {
  cleanup();
});

describe('Molecules/TaskItem', () => {
  test('正しくレンダリングされている', () => {
    const element = renderer
      .create(
        <MemoryRouter>
          <TaskItem
            id={id}
            title={title}
            description={description}
            isCompleted={isCompleted}
            priority={priority}
            completedStateHandler={completedStateHandler}
            priorityStateHandler={priorityStateHandler}
          />
        </MemoryRouter>,
      )
      .toJSON();
    expect(element).toMatchSnapshot();
  });

  test('チェックボタンを押下できる', () => {
    render(
      <MemoryRouter>
        <TaskItem
          id={id}
          title={title}
          description={description}
          isCompleted={isCompleted}
          priority={priority}
          completedStateHandler={completedStateHandler}
          priorityStateHandler={priorityStateHandler}
        />
      </MemoryRouter>,
    );
    userEvent.click(screen.getByRole('checkbox'));
    expect(completedStateHandler).toHaveBeenCalled();
  });

  test('スターボタンを押下できる', () => {
    render(
      <MemoryRouter>
        <TaskItem
          id={id}
          title={title}
          description={description}
          isCompleted={isCompleted}
          priority={priority}
          completedStateHandler={completedStateHandler}
          priorityStateHandler={priorityStateHandler}
        />
      </MemoryRouter>,
    );
    userEvent.click(screen.getByTestId('star'));
    expect(priorityStateHandler).toHaveBeenCalled();
  });
});
