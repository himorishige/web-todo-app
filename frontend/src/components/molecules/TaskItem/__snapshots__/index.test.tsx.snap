// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/TaskItem 正しくレンダリングされている 1`] = `
<div
  className="css-dbsdw7-TaskItem"
>
  <div
    className="css-1o4cpu6-TaskItem"
  >
    <label
      className="css-dygvpw"
    >
      <input
        className="input"
        defaultChecked={false}
        disabled={false}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="dummy-input"
        data-testid="checkbutton"
      />
    </label>
    <div
      className="css-1fmwtpk-TaskItem"
    >
      <a
        href="/tasks/1"
        onClick={[Function]}
      >
        タイトル
        <svg
          className="css-42jkzv-TaskItem"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.502 1.94a.5.5 0 010 .706L14.459 3.69l-2-2L13.502.646a.5.5 0 01.707 0l1.293 1.293zm-1.75 2.456l-2-2L4.939 9.21a.5.5 0 00-.121.196l-.805 2.414a.25.25 0 00.316.316l2.414-.805a.5.5 0 00.196-.12l6.813-6.814z"
          />
          <path
            clipRule="evenodd"
            d="M1 13.5A1.5 1.5 0 002.5 15h11a1.5 1.5 0 001.5-1.5v-6a.5.5 0 00-1 0v6a.5.5 0 01-.5.5h-11a.5.5 0 01-.5-.5v-11a.5.5 0 01.5-.5H9a.5.5 0 000-1H2.5A1.5 1.5 0 001 2.5v11z"
            fillRule="evenodd"
          />
        </svg>
      </a>
    </div>
  </div>
  <div>
    <div
      className="css-kknodv"
      onClick={[Function]}
    >
      <svg
        className="css-ba990d"
        data-testid="star"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M2.866 14.85c-.078.444.36.791.746.593l4.39-2.256 4.389 2.256c.386.198.824-.149.746-.592l-.83-4.73 3.523-3.356c.329-.314.158-.888-.283-.95l-4.898-.696L8.465.792a.513.513 0 00-.927 0L5.354 5.12l-4.898.696c-.441.062-.612.636-.283.95l3.523 3.356-.83 4.73zm4.905-2.767l-3.686 1.894.694-3.957a.565.565 0 00-.163-.505L1.71 6.745l4.052-.576a.525.525 0 00.393-.288l1.847-3.658 1.846 3.658a.525.525 0 00.393.288l4.052.575-2.906 2.77a.564.564 0 00-.163.506l.694 3.957-3.686-1.894a.503.503 0 00-.461 0z"
          fillRule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;
